### PR TITLE
EZP-26391: ez-support-tools factory called on every request

### PR DESCRIPTION
**Pull request by glye**
> Fixes https://jira.ez.no/browse/EZP-26391
> Part 1: https://github.com/ezsystems/ez-support-tools/pull/20
> Status: QA

Lazy loading was failing because ezcSystemInfo uses magic properties, while the proxy manager looks for public properties. This fixes the issue by wrapping ezcSystemInfo and exposing the magic properties as public ones.

Manual test: Verified that both the web interface and the command work as expected, and that the factory build method is only called when needed.

### CODE REVIEW

**Comment by bdunogier** [comment]
Looks okay.

What bothers me a tiny bit is that there _is_ code in the class generated for the lazy proxy to handle `__get()`, and I'm not really sure why it doesn't work in dev, but does in prod. Probably missing something, but it's not really code that is meant to be read.

**Comment by glye** [comment]
Fixing failing tests.

**Comment by glye** [comment]
Ping @andrerom @yannickroger - this fixes the problem QA found in https://github.com/ezsystems/ez-support-tools/pull/20

**Comment by andrerom** [comment]
Did you by any chance look at if we can use Proxy manager 2.x? It might solve the issue.
https://github.com/Ocramius/ProxyManager/releases

**Comment by bdunogier** [comment]
I haven't yet. Let's create a story for this.

**Comment by andrerom** [comment]
it requires php 7, ignore..

**Comment by bdunogier** [comment]
Frak :(

**Comment by glye** [comment]
Ack. Time may be too short to get this fix in by release, I'll un-lazy the service later today if we don't make it, and re-do the PR.

File: SystemInfo/EzcSystemInfoWrapper.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    /** @var \ezcSystemInfoAccelerator */
+    public $phpAccelerator;
+
+    /** @var bool */
+    public $isShellExecution;
+
+    public function __construct()
+    {
+        $ezcSystemInfo = ezcSystemInfo::getInstance();
```
Comment (andrerom) [inline SystemInfo/EzcSystemInfoWrapper.php]: Could you please also take care about the exception while on it, as the upstream pull is taking a long time, ref:
```php
        try {
            $ezcSystemInfo = ezcSystemInfo::getInstance();
        } catch(ezcSystemInfoReaderCantScanOSException $e) {
            // Leave properties as null: https://github.com/zetacomponents/SystemInformation/pull/9
            return;
        }
```

**Comment by glye** [comment]
Making the service non-lazy for now: https://github.com/ezsystems/ez-support-tools/pull/22
but I'll resolve this.

File: SystemInfo/EzcSystemInfoFactory.php [hunk 2]
```diff
@@ -8,15 +8,13 @@
...
-
 /**
  * Factory for zetacomponents/sysinfo. Used here to simplify testing.
  */
 class EzcSystemInfoFactory
 {
     public static function buildEzcSystemInfo()
     {
-        return ezcSystemInfo::getInstance();
+        return new EzcSystemInfoWrapper();
```
Comment (glye) [inline SystemInfo/EzcSystemInfoFactory.php]: Affirmative.

**Comment by glye** [comment]
Update: Removed the redundant factory, and catch `ezcSystemInfoReaderCantScanOSException`.

**Comment by andrerom** [comment]
Nice, wdt @bdunogier ? Seems like the way to go imoh over #22, as this will also fix issues people have with blocked usage when ezcSystemInfoReader is checked and can't read system info, #22 won't afaik.

**Comment by glye** [comment]
> #22 will have that issue afaik, as it's not lazy and thus will be executed all the time

True.

**Comment by bdunogier** [comment]
I indeed prefer this one too. Could QA test it ? I'll have another look and hopefully approve it.

File: SystemInfo/EzcSystemInfoWrapper.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    public function __construct()
+    {
+        try {
+            $ezcSystemInfo = ezcSystemInfo::getInstance();
+        } catch(ezcSystemInfoReaderCantScanOSException $e) {
+            // Leave properties as null: https://github.com/zetacomponents/SystemInformation/pull/9
+            return;
+        }
+
+        $this->osType = $ezcSystemInfo->osType;
```
Comment (bdunogier) [inline SystemInfo/EzcSystemInfoWrapper.php]: You could probably shorten that to some kind of loop:

```
foreach (get_object_vars($this) as $var) {
  $this->$var = $ezcSystemInfo->$var;
}
```

**Comment by bdunogier** [comment]
Let's send that to QA asap. The issue is annoying enough, and it'd be a shame not to fix it. Worst thing that happen, we'll do another support tools RC :)

File: SystemInfo/EzcSystemInfoWrapper.php (hunk 3 above)
Comment (andrerom) [inline SystemInfo/EzcSystemInfoWrapper.php]: well, it also makes it less readable, so if QA approves it like this lets then maybe skip changing this part if that is ok.

Comment (bdunogier) [inline SystemInfo/EzcSystemInfoWrapper.php]: No strong feelings :)

**Comment by glye** [comment]
Update: Shortened wrapper property assignment.

**Comment by pcardiga** [comment]
PR verified by QA

**Comment by glye** [comment]
Thanks! Just waiting for Travis then...

**Pull request merged**